### PR TITLE
Merge dynamic_endpoint_extensions into master

### PR DESCRIPTION
**Pull request by Captain-P-Goldfish**
This will merge the dynamic_endpoint_extensions branch into the master branch.
Additionally the error-handling within the bulk-endpoint was fixed

### CODE REVIEW

**Comment by Captain-P-Goldfish** [comment]
Hello would anybody please say anything on this?

**Comment by Captain-P-Goldfish** [comment]
ping...
please at least give me any answer if you will merge this into master or not. I need to know this so that we can plan our next steps. If you will not merge this into master we will need to think of a solution very soon.
Please do not stay silent...

File: modules/charon-core/pom.xml [hunk 1]
```diff
@@ -34,28 +34,38 @@
...
+                </exclusion>
+            </exclusions>
         </dependency>
+
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <scope>test</scope>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
```
Comment (DMHP) [inline modules/charon-core/pom.xml]: Why did we need commons lang3?

File: modules/charon-core/pom.xml [hunk 2]
```diff
@@ -34,28 +34,38 @@
...
+                <exclusion>
+                    <artifactId>commons-logging</artifactId>
+                    <groupId>commons-logging</groupId>
+                </exclusion>
+            </exclusions>
         </dependency>
         <dependency>
             <groupId>org.apache.ws.commons.axiom</groupId>
             <artifactId>axiom-impl</artifactId>
+            <exclusions>
```
Comment (DMHP) [inline modules/charon-core/pom.xml]: commons-logging is excluded two times. Any specific reason why did you remove this dependency?

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/ComplexAttribute.java [hunk 3]
```diff
@@ -58,7 +60,7 @@ public void setSubAttributesList(Map<String, Attribute> subAttributesList) {
      * @param attributeName
      * @return Attribute
      */
-    public Attribute getSubAttribute(String attributeName) throws CharonException {
+    public Attribute getSubAttribute(String attributeName) {
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/ComplexAttribute.java]: We are not recommending to change public methods as we need to preserve the backward compatibility as these public methods can be used from different other places.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/ComplexAttribute.java [hunk 4]
```diff
@@ -71,10 +73,22 @@ public Attribute getSubAttribute(String attributeName) throws CharonException {
      * @throws CharonException
      */
     @Override
-    public void deleteSubAttributes() throws CharonException {
+    public void deleteSubAttributes() {
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/ComplexAttribute.java]: We are not recommending to change public methods as we need to preserve the backward compatibility as these public methods can be used from different other places.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/DefaultAttributeFactory.java [hunk 5]
```diff
@@ -28,15 +29,17 @@
...
     /*
      * Returns the defined type of attribute with the user defined value
      * included and necessary attribute characteristics set
      * @param attributeSchema - Attribute schema
      * @param attribute - attribute
      * @return Attribute
      */
-    public static Attribute createAttribute(AttributeSchema attributeSchema,
-                                            AbstractAttribute attribute) throws CharonException, BadRequestException {
+    public static Attribute createAttribute(AttributeSchema attributeSchema, AbstractAttribute attribute)
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/DefaultAttributeFactory.java]: We are not recommending to change public methods as we need to preserve the backward compatibility as these public methods can be used from different other places.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/DefaultAttributeFactory.java [hunk 6]
```diff
@@ -90,17 +90,16 @@ public static Attribute createAttribute(AttributeSchema attributeSchema,
...
      *
      * @param attributeValue
      * @param attributeDataType
      * @return boolean
      * @throws BadRequestException
      */
-    protected static boolean isAttributeDataTypeValid(Object attributeValue,
-                                                      SCIMDefinitions.DataType attributeDataType)
-                                                throws BadRequestException {
+    protected static boolean isAttributeDataTypeValid(Object attributeValue, SCIMDefinitions.DataType attributeDataType)
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/DefaultAttributeFactory.java]: We are not recommending to change public methods as we need to preserve the backward compatibility as these public methods can be used from different other places.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/SimpleAttribute.java [hunk 7]
```diff
@@ -125,12 +122,12 @@ public Boolean getBooleanValue() throws CharonException {
...
      * uodate the attribute value
-     * 
+     *
      * @param value
-     * 
+     *
      * @throws CharonException
      */
-    public void updateValue(Object value) throws CharonException {
+    public void updateValue(Object value) {
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/SimpleAttribute.java]: We are not recommending to change public methods as we need to preserve the backward compatibility as these public methods can be used from different other places.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/AuthenticationScheme.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.Objects;
+
+/**
+ * A multi-valued complex type that specifies supported
+ * authentication scheme properties.  To enable seamless discovery of
+ * configurations, the service provider SHOULD, with the appropriate
+ * security considerations, make the authenticationSchemes attribute
+ * publicly accessible without prior authentication.  REQUIRED.
+ * <br><br>
+ * created at: 17.04.2019
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/AuthenticationScheme.java]: Please add wso2 license header

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/BulkFeature.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
+package org.wso2.charon3.core.config;
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/BulkFeature.java]: Please add wso2 license header

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/FilterFeature.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package org.wso2.charon3.core.config;
+
+import java.util.Objects;
+
+/**
+ * A complex type that specifies FILTER options.  REQUIRED.  See Section 3.4.2.2 of [RFC7644].
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/FilterFeature.java]: Please add WSO2 License header

File: modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/UserManager.java [hunk 11]
```diff
@@ -1,143 +0,0 @@
...
-
-import java.util.List;
-import java.util.Map;
-
-/**
- * This is the interface for usermanager extension.
- * An implementation can plugin their own user manager-(either LDAP based, DB based etc)
- * by implementing this interface and mentioning it in configuration.
- */
-public interface UserManager {
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/extensions/UserManager.java]: This is a main interface in Charon repo. Removing public interfaces are not recommended.  There may be users who use this class. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/ResourceTypeRegistration.java [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+import org.wso2.charon3.core.schema.SCIMResourceTypeExtensionSchema;
+import org.wso2.charon3.core.schema.SCIMSchemaDefinitions;
+import org.wso2.charon3.core.utils.LambdaExceptionUtils;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * this class represents a configuration class that is used to register resource types in charon. All registered
```
Comment (DMHP) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/ResourceTypeRegistration.java]: Please add WSO2 license header

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/BulkRequestProcessor.java [hunk 13]
```diff
@@ -1,201 +0,0 @@
...
-import org.wso2.charon3.core.objects.bulk.BulkResponseData;
-import org.wso2.charon3.core.protocol.endpoints.GroupResourceManager;
-import org.wso2.charon3.core.protocol.endpoints.ResourceManager;
-import org.wso2.charon3.core.protocol.endpoints.UserResourceManager;
-import org.wso2.charon3.core.schema.SCIMConstants;
-
-/**
- *
- */
-public class BulkRequestProcessor {
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/BulkRequestProcessor.java]: Removing public classes are not recommended.  There may be users who use this class. Changes should be done in a backward-compatible manner.

File: modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java [hunk 14]
```diff
@@ -41,16 +40,14 @@
...
+
     //in memory user manager stores users
     ConcurrentHashMap<String, User> inMemoryUserList = new ConcurrentHashMap<String, User>();
-    ConcurrentHashMap<String, Group> inMemoryGroupList = new ConcurrentHashMap<String, Group>();
 
 
     @Override
-    public User createUser(User user, Map<String, Boolean> map)
-            throws CharonException, ConflictException, BadRequestException {
+    public User create(User user, Map<String, Boolean> map) throws AbstractCharonException {
```
Comment (DMHP) [inline modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java]: It is not recommended to change public method signatures as they can be used from some other places as well. We need to preserve the backward compatibility.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/GroupResourceManager.java [hunk 15]
```diff
@@ -1,770 +0,0 @@
...
-import java.util.Map;
-
-
-/**
- * REST API exposed by Charon-Core to perform operations on GroupResource.
- * Any SCIM service provider can call this API perform relevant CRUD operations on Group ,
- * based on the HTTP requests received by SCIM Client.
- */
-
-public class GroupResourceManager extends AbstractResourceManager {
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/GroupResourceManager.java]: This is a main class in Charon repo. Removing public classes are not recommended.  There may be users who use this class. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/MeResourceManager.java [hunk 16]
```diff
@@ -1,424 +0,0 @@
...
-import java.util.List;
-import java.util.Map;
-
-/**
- * A client MAY use a URL of the form "<base-uri>/Me" as a uri alias for
- * the User or other resource associated with the currently
- * authenticated subject for any SCIM operation.
- */
-
-public class MeResourceManager extends AbstractResourceManager {
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/MeResourceManager.java]: This is a main class in Charon repo. Removing public classes are not recommended.  There may be users who use this class. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/ResourceTypeResourceManager.java [hunk 17]
```diff
@@ -15,213 +15,188 @@
...
+    public SCIMResponse delete(String id) {
 
         String error = "Request is undefined";
         BadRequestException badRequestException = new BadRequestException(error, ResponseCodeConstants.INVALID_PATH);
         return encodeSCIMException(badRequestException);
     }
 
     @Override
-    public SCIMResponse listWithGET(UserManager userManager, String filter, int startIndex, int count, String sortBy,
+    public SCIMResponse listWithGET(String filter, Integer startIndexInt, Integer countInt, String sortBy,
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/ResourceTypeResourceManager.java]: Changing public APIs are not recommended.  There may be users who use this. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/ResourceTypeResourceManager.java [hunk 18]
```diff
@@ -15,213 +15,188 @@
...
+            log.error(ex.getMessage(), ex);
+            CharonException charonException = new CharonException("an unexpected error occured: " + ex.getMessage());
+            return encodeSCIMException(charonException);
         }
     }
 
     @Override
-    public SCIMResponse create(String scimObjectString, UserManager userManager, String attributes, String
-            excludeAttributes) {
+    public SCIMResponse create(String scimObjectString, String attributes, String excludeAttributes) {
```
Comment (IsuraD) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/protocol/endpoints/ResourceTypeResourceManager.java]: Changing public APIs are not recommended.  There may be users who use this. Changes should be done in a backward-compatible manner.

File: modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java [hunk 19]
```diff
@@ -60,18 +57,16 @@ public User createUser(User user, Map<String, Boolean> map)
     }
 
     @Override
-    public User getUser(String id, Map<String, Boolean> map)
```
Comment (DMHP) [inline modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java]: Seems like you have removed UserManager Interface and added ResourceHandler interface. We are not recommending to remove the public interfaces as we need to preserve the backward compatibility as the other users may depend on those interfaces.

File: modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java [hunk 20]
```diff
@@ -60,18 +57,16 @@ public User createUser(User user, Map<String, Boolean> map)
...
+            return (User) CopyUtil.deepCopy(inMemoryUserList.get(id));
+        } else {
+            throw new NotFoundException("No user with the id : " + id);
+        }
     }
 
     @Override
-    public void deleteUser(String id)
-            throws NotFoundException, CharonException, NotImplementedException, BadRequestException {
+    public void delete(String id) throws NotFoundException {
```
Comment (DMHP) [inline modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java]: It is not recommended to change public method signatures as they can be used from some other places as well. We need to preserve the backward compatibility.

File: modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java [hunk 21]
```diff
@@ -80,12 +75,16 @@ public void deleteUser(String id)
     }
 
     @Override
-    public List<Object> listUsersWithGET(Node rootNode, int startIndex, int count, String sortBy,
-                                         String sortOrder, String domainName, Map<String, Boolean> requiredAttributes)
-            throws CharonException, NotImplementedException, BadRequestException {
+    public List<Object> listResources(Node rootNode,
```
Comment (DMHP) [inline modules/charon-utils/src/main/java/org/wso2/charon3/utils/usermanager/InMemoryUserManager.java]: It is not recommended to change public method signatures as they can be used from some other places as well. We need to preserve the backward compatibility.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/MultiValuedAttribute.java [hunk 22]
```diff
@@ -67,23 +69,23 @@ public Attribute getSubAttribute(String attributeName) throws CharonException {
      * @throws CharonException
      */
     @Override
-    public void deleteSubAttributes() throws CharonException {
+    public void deleteSubAttributes() {
```
Comment (ashensw) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/MultiValuedAttribute.java]: Changing public method signatures is not recommended. There may be users who use this method already. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/MultiValuedAttribute.java [hunk 23]
```diff
@@ -67,23 +69,23 @@ public Attribute getSubAttribute(String attributeName) throws CharonException {
...
     }
 
-     /*
+    /*
      * clear all primitive values
      * @throws CharonException
      */
 
-    public void deletePrimitiveValues() throws CharonException {
+    public void deletePrimitiveValues() {
```
Comment (ashensw) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/attributes/MultiValuedAttribute.java]: Changing public method signatures is not recommended. There may be users who use this method already. Changes should be done in a backward-compatible manner.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/SchemaRegistration.java [hunk 24]
```diff
@@ -0,0 +1,82 @@
+package org.wso2.charon3.core.config;
```
Comment (ashensw) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/SchemaRegistration.java]: Please add the WSO2 License header.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/ScimConfiguration.java [hunk 25]
```diff
@@ -0,0 +1,410 @@
+package org.wso2.charon3.core.config;
```
Comment (ashensw) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/ScimConfiguration.java]: Please add the WSO2 License header.

File: modules/charon-core/src/main/java/org/wso2/charon3/core/config/ScimFeature.java [hunk 26]
```diff
@@ -0,0 +1,56 @@
+package org.wso2.charon3.core.config;
```
Comment (ashensw) [inline modules/charon-core/src/main/java/org/wso2/charon3/core/config/ScimFeature.java]: Please add the WSO2 License header.

**Comment by tharindu-bandara** [comment]
Hi @Captain-P-Goldfish,

Were you able to check on the comments provided above? As this PR includes significant API changes, we would need to resolve all the provided comments before merging.

**Comment by Captain-P-Goldfish** [comment]
I am going to close this pull request. Since you saw that my changes are quiet significant I am not able to serve all of the required comments. 
Besides I used the waiting time on this issue to develop my own SCIM open source project at https://github.com/Captain-P-Goldfish/SCIM-SDK. The server side is also feature complete and dynamically extendable
